File: packages/types-ts/events.ts
import type { z } from 'zod';
import { RecipeParsedEventSchema, ScrapeRequestedEventSchema } from './schema';
import { Topics } from './topics';

// Type resolution
export type ScrapeRequestEvent = z.infer<typeof ScrapeRequestedEventSchema>;
export type RecipeParsedEvent = z.infer<typeof RecipeParsedEventSchema>;

// Topic mapping
export const EventSchemasByTopic = {
  [Topics.RECIPE_PARSED]: RecipeParsedEventSchema,
  [Topics.SCRAPE_REQUESTED]: ScrapeRequestedEventSchema,
} as const;

// derive the payload type for each topic directly from its Zod schema
export type EventPayloadMap = {
  [K in keyof typeof EventSchemasByTopic]: z.infer<typeof EventSchemasByTopic[K]>
};

File: packages/py-utils/src/py_utils/events.py
from py_utils.schema.RecipeParsedEvent import RecipeParsedEvent
from py_utils.schema.ScrapeRequestedEvent import ScrapeRequestedEvent
from py_utils.topics import TOPICS

EVENT_TOPICS_BY_SCHEMA = {
  TOPICS["RECIPE_PARSED"]: RecipeParsedEvent,
  TOPICS["SCRAPE_REQUESTED"]: ScrapeRequestedEvent
}

